eth/eth2.go

```diff
@@ -0,0 +1,325 @@
+	"github.com/ethereum/go-ethereum/log"
+	chainParams "github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+type Eth2API struct {
+	eth  *Ethereum
+	env  *eth2bpenv
+	head common.Hash
+}
+
+// NewEth2API creates a new API definition for the eth2 prototype.
+func NewEth2API(eth *Ethereum) *Eth2API {
+	return &Eth2API{eth: eth}
+}
+
+type eth2bpenv struct {
+	state   *state.StateDB
+	tcount  int
+	gasPool *core.GasPool
+
+	header   *types.Header
+	txs      []*types.Transaction
+	receipts []*types.Receipt
+}
+
+func (api *Eth2API) commitTransaction(tx *types.Transaction, coinbase common.Address) error {
+	//snap := eth2rpc.current.state.Snapshot()
+
+	chain := api.eth.BlockChain()
+	receipt, err := core.ApplyTransaction(chain.Config(), chain, &coinbase, api.env.gasPool, api.env.state, api.env.header, tx, &api.env.header.GasUsed, *chain.GetVMConfig())
+	if err != nil {
+		//w.current.state.RevertToSnapshot(snap)
+		return err
+	}
+	api.env.txs = append(api.env.txs, tx)
+	api.env.receipts = append(api.env.receipts, receipt)
+
+	return nil
+}
+
+func (api *Eth2API) makeEnv(parent *types.Block, header *types.Header) error {
+	state, err := api.eth.BlockChain().StateAt(parent.Root())
+	if err != nil {
+		return err
+	}
+	api.env = &eth2bpenv{
+		state:   state,
+		header:  header,
+		gasPool: new(core.GasPool).AddGas(header.GasLimit),
+	}
+	return nil
+}
+
+// Structure described at https://hackmd.io/T9x2mMA4S7us8tJwEB3FDQ
+type AssembleBlockParams struct {
+	ParentHash common.Hash `json:"parent_hash"`
+	Timestamp  uint64      `json:"timestamp"`
+}
+
+// Structure described at https://ethresear.ch/t/executable-beacon-chain/8271
+type ExecutableData struct {
+	Coinbase     common.Address       `json:"coinbase"`
+	StateRoot    common.Hash          `json:"state_root"`
+	GasLimit     uint64               `json:"gas_limit"`
+	GasUsed      uint64               `json:"gas_used"`
+	Transactions []*types.Transaction `json:"transactions"`
+	ReceiptRoot  common.Hash          `json:"receipt_root"`
+	LogsBloom    []byte               `json:"logs_bloom"`
+	BlockHash    common.Hash          `json:"block_hash"`
+	ParentHash   common.Hash          `json:"parent_hash"`
+	Difficulty   *big.Int             `json:"difficulty"`
+}
+
+// AssembleBlock creates a new block, inserts it into the chain, and returns the "execution
+// data" required for eth2 clients to process the new block.
+func (api *Eth2API) AssembleBlock(params AssembleBlockParams) (*ExecutableData, error) {
+	log.Info("Produce block", "parentHash", params.ParentHash)
+
+	bc := api.eth.BlockChain()
+	parent := bc.GetBlockByHash(params.ParentHash)
+	pool := api.eth.TxPool()
+
+	if parent.Time() >= params.Timestamp {
+		return nil, fmt.Errorf("child timestamp lower than parent's: %d >= %d", parent.Time(), params.Timestamp)
+	}
+	// this will ensure we're not going off too far in the future
+	if now := uint64(time.Now().Unix()); params.Timestamp > now+1 {
```

Comment:
I'm not sure the sanity check is needed here. I image it's possible in the future that users put beacon node and geth node into two separate machines and connect them via RPC. So it's possible that the geth node machine is time delayed and return the block wait long time sleeping(e.g. 5 second).

IIUC, in the eth2, the block has to be purposed in the given time limit. Maybe it would be more sense to do the sleeping in the eth2 part?

Just throw out the thoughts.